pkg/v1workermeta/meta.go

```diff
@@ -0,0 +1,161 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1workermeta
+
+import (
+	"github.com/golang/protobuf/proto"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/util"
+
+	"github.com/pingcap/dm/dm/pb"
+	"github.com/pingcap/dm/pkg/helper"
+	"github.com/pingcap/dm/pkg/terror"
+)
+
+// meta stores metadata of tasks.
+type meta struct {
+	tasks map[string]*pb.V1SubTaskMeta
+}
+
+// newMeta returns a meta object.
+func newMeta(db *leveldb.DB) (*meta, error) {
+	tasks, err := loadTaskMetas(db)
+	if err != nil {
+		return nil, err
+	}
+
+	return &meta{
+		tasks: tasks,
+	}, nil
+}
+
+// TasksMeta returns meta of all tasks.
+func (meta *meta) TasksMeta() map[string]*pb.V1SubTaskMeta {
+	tasks := make(map[string]*pb.V1SubTaskMeta, len(meta.tasks))
+	for name, task := range meta.tasks {
+		tasks[name] = proto.Clone(task).(*pb.V1SubTaskMeta)
+	}
+	return tasks
+}
+
+// taskMetaPrefix is prefix of task meta key.
+var taskMetaPrefix = []byte("!DM!TaskMeta")
+
+// encodeTaskMetaKey encodes take name into a task meta key.
+func encodeTaskMetaKey(name string) []byte {
+	key := append([]byte{}, taskMetaPrefix...)
+	return append(key, name...)
+}
+
+// loadTaskMetas loads all task metas from kv db.
+func loadTaskMetas(db *leveldb.DB) (map[string]*pb.V1SubTaskMeta, error) {
+	if helper.IsNil(db) {
+		return nil, terror.ErrWorkerLogInvalidHandler.Generate()
+	}
+
+	var (
+		tasks = make(map[string]*pb.V1SubTaskMeta)
+		err   error
+	)
+
+	iter := db.NewIterator(util.BytesPrefix(taskMetaPrefix), nil)
+	for iter.Next() {
+		taskBytes := iter.Value()
+		task := &pb.V1SubTaskMeta{}
+		err = task.Unmarshal(taskBytes)
+		if err != nil {
+			err = terror.ErrWorkerLogUnmarshalTaskMeta.Delegate(err, taskBytes)
+			break
+		}
+
+		tasks[task.Name] = task
+	}
+	iter.Release()
+	if err != nil {
+		return nil, terror.ErrWorkerLogFetchTaskFromMeta.Delegate(err, taskMetaPrefix)
+	}
+
+	err = iter.Error()
+	if err != nil {
+		return nil, terror.ErrWorkerLogFetchTaskFromMeta.Delegate(err, taskMetaPrefix)
+	}
+
+	return tasks, nil
+}
+
+// setTaskMeta saves task meta into kv db.
+// it's used for testing.
+func setTaskMeta(db *leveldb.DB, task *pb.V1SubTaskMeta) error {
+	if helper.IsNil(db) {
+		return terror.ErrWorkerLogInvalidHandler.Generate()
+	}
+
+	err := verifyTaskMeta(task)
+	if err != nil {
+		return terror.Annotatef(err, "verify task meta %+v", task)
+	}
+
+	taskBytes, err := task.Marshal()
+	if err != nil {
+		return terror.ErrWorkerLogMarshalTask.Delegate(err, task)
+	}
+
+	err = db.Put(encodeTaskMetaKey(task.Name), taskBytes, nil)
+	if err != nil {
+		return terror.ErrWorkerLogSaveTaskMeta.Delegate(err, task.Name)
+	}
+
+	return nil
+}
+
+// deleteTaskMeta deletes task meta from kv DB.
+// it's used for testing.
+func deleteTaskMeta(db *leveldb.DB, name string) error {
+	if helper.IsNil(db) {
```

Comment:
ditto

Reply:
changed to `db == nil` in https://github.com/pingcap/dm/pull/846/commits/042b647f3d97e76cca0e6318e0d1b7b1b769d75b.